### PR TITLE
Migrate backup sync controller from code-generator to kubebuilder

### PR DESCRIPTION
**Pull request by blackpiglet**
1. use kubebuilder's reconcile logic to replace controller's old logic.
2. use ginkgo and gomega to replace testing.

Signed-off-by: Xun Jiang <jxun@vmware.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required`.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/cmd/server/server.go [hunk 1]
```diff
@@ -769,7 +747,7 @@ func (s *server) runControllers(defaultVolumeSnapshotLocations map[string]string
 	}
 
 	enabledControllers := map[string]func() controllerRunInfo{
-		controller.BackupSync:        backupSyncControllerRunInfo,
+		//controller.BackupSync:        backupSyncControllerRunInfo,
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/cmd/server/server.go:750]: Thanks. Removed.

File: pkg/controller/backup_sync_controller.go [hunk 2]
```diff
@@ -38,114 +36,50 @@ import (
...
-}
-
-func (c *backupSyncController) run() {
-	c.logger.Debug("Checking for existing backup storage locations to sync into cluster")
+func (b *BackupSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := b.Logger.WithField("controller", BackupSync)
+	log.Debug("Checking for existing backup storage locations to sync into cluster.")
 
-	locationList, err := storage.ListBackupStorageLocations(context.Background(), c.kbClient, c.namespace)
+	locationList, err := storage.ListBackupStorageLocations(ctx, b.Client, b.Namespace)
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller.go:61]: Great catch. It should be b.Ctx.

File: pkg/controller/backup_sync_controller.go [hunk 3]
```diff
@@ -303,36 +237,45 @@ func (c *backupSyncController) run() {
...
 		statusPatch := client.MergeFrom(location.DeepCopy())
 		location.Status.LastSyncedTime = &metav1.Time{Time: time.Now().UTC()}
-		if err := c.kbClient.Status().Patch(context.Background(), &location, statusPatch); err != nil {
+		if err := b.Client.Status().Patch(b.Ctx, &location, statusPatch); err != nil {
 			log.WithError(errors.WithStack(err)).Error("Error patching backup location's last-synced time")
 			continue
 		}
 	}
+
+	return ctrl.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller.go:257]: Thanks. I missed this part of previous version of controller.
c.resyncPeriod = 30 * time.Second

File: pkg/controller/backup_sync_controller.go [hunk 4]
```diff
@@ -38,114 +36,50 @@ import (
...
-	namespace string,
-	csiSnapshotClient *snapshotterClientSet.Clientset,
-	kubeClient kubernetes.Interface,
-	defaultBackupLocation string,
-	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
-	backupStoreGetter persistence.ObjectBackupStoreGetter,
-	logger logrus.FieldLogger,
-) Interface {
-	if syncPeriod <= 0 {
-		syncPeriod = time.Minute
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller.go:75]: I totally missed this part. I just treated this function as boilerplate.
Thanks a lot. I will add the syncPeriod setting logic back in Reconcile function.

File: pkg/controller/backup_sync_controller.go [hunk 5]
```diff
@@ -38,114 +36,50 @@ import (
...
-	newPluginManager        func(logrus.FieldLogger) clientmgmt.Manager
-	backupStoreGetter       persistence.ObjectBackupStoreGetter
+type BackupSyncReconciler struct {
+	Ctx                     context.Context
+	Client                  client.Client
+	PodVolumeBackupClient   velerov1client.PodVolumeBackupsGetter
+	BackupLister            velerov1listers.BackupLister
+	BackupClient            velerov1client.BackupsGetter
+	Namespace               string
+	DefaultBackupLocation   string
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller.go:50]: Thanks. I will add code to iterate the BSL list

File: pkg/controller/backup_sync_controller_test.go [hunk 6]
```diff
@@ -107,263 +113,282 @@ func defaultLocationsListWithLongerLocationName(namespace string) []*velerov1api
...
+				Client:                  ctrlfake.NewClientBuilder().Build(),
+				BackupClient:            client.VeleroV1(),
+				PodVolumeBackupClient:   client.VeleroV1(),
+				BackupLister:            sharedInformers.Velero().V1().Backups().Lister(),
+				Namespace:               test.namespace,
+				DefaultBackupSyncPeriod: time.Second * 10,
+				NewPluginManager:        func(logrus.FieldLogger) clientmgmt.Manager { return pluginManager },
+				BackupStoreGetter:       NewFakeObjectBackupStoreGetter(backupStores),
+				Logger:                  velerotest.NewLogger(),
+				//DefaultBackupLocation:   "location-1",
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller_test.go:381]: Removed.

File: pkg/controller/backup_sync_controller_test.go [hunk 7]
```diff
@@ -107,263 +113,282 @@ func defaultLocationsListWithLongerLocationName(namespace string) []*velerov1api
...
-				"",
-				func(logrus.FieldLogger) clientmgmt.Manager { return pluginManager },
-				NewFakeObjectBackupStoreGetter(backupStores),
-				velerotest.NewLogger(),
-			).(*backupSyncController)
-
 			pluginManager.On("CleanupClients").Return(nil)
+			r := BackupSyncReconciler{
+				Ctx: ctx,
+				//Client:                  ctrlfake.NewFakeClientWithScheme(scheme.Scheme),
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/controller/backup_sync_controller_test.go:371]: Removed.

**Review by zubron — APPROVED** [review]
LGTM - thanks!

**Review by reasonerjt — APPROVED** [review]

**Review by zubron — APPROVED** [review]

**Pull request merged**